[WC-884] Rewrite Range slider to pluggable

iobuhov opened the pull request:
## Checklist
- Contains unit tests ❌
- Contains breaking changes ✅
- Contains Atlas changes ✅
- Compatible with: MX 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ✅
- Compatible with Studio ✅
- Cross-browser compatible ✅

#### Feature specific
- Comply with designs ✅
- Comply with PM's requirements ✅

## This PR contains
- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Convert RangeSlider to PIW

## Relevant changes
- Moing `range-slider` to `pluggableWidgets`
- Update `package.json`
- Introduce new properties

## What should be covered while testing?
As user you should be able to change model attributes using this widget

## Extra comments (optional)

Commit:
refactor(range-slider-web): move range-slider to pluggable widgets

Commit:
feat(range-slider-web): add initial version

Commit:
feat(range-slider-web): convert range-slider to new format

Commit:
feat(atlas-core): add design properties for range-slider

Commit:
feat(range-slider-web): add uinit tests

Keraito commented on packages/pluggableWidgets/range-slider-web/src/utils/isVertical.ts:
```diff
@@ -0,0 +1,5 @@
+import { RangeSliderContainerProps } from "../../typings/RangeSliderProps";
+
+export function isVertical(params: Pick<RangeSliderContainerProps, "orientation">): boolean {
+    return params.orientation === "vertical";
+}
```
Nit: do we really need this is as a separate util instead of just inline ` === "vertcal"`?

iobuhov replied:
We don't, but this is left over and I think this is fine leave this as it is for now

Keraito commented on packages/theming/atlas/src/themesource/atlas_core/web/core/widgets/_range-slider.scss:
```diff
@@ -14,108 +15,16 @@
-            touch-action: none;
-            -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
-        }
-
-        .rc-slider-rail {
-            position: absolute;
-            width: 100%;
-            height: 4px;
-            border-radius: 6px;
-            background-color: #e9e9e9;
-        }
-
-        .rc-slider-track {
-            position: absolute;
-            left: 0;
-            height: 4px;
-            border-radius: 6px;
-        }
-
-        .rc-slider-step {
-            position: absolute;
-            width: 100%;
-            height: 4px;
-            background: transparent;
-        }
-
-        .rc-slider-handle,
-        .rc-slider-dot-active {
-            position: absolute;
-            width: 14px;
-            height: 14px;
-            margin-top: -5px;
-            margin-left: -7px;
-            cursor: grab;
-            border: 2px solid $brand-default;
-            border-radius: 50%;
-            background-color: #ffffff;
-            touch-action: pan-x;
-
-            &:focus {
```
Is it fine to remove everything due to potential backwards compatibility?

iobuhov replied:
This code migrated to atlas_core. It's just styling, nothing more.

Keraito replied:
Even if they're migrated towards atlas_core, we still need to keep them around as far as I remember. But if you did something similar to Slider, then it's fine cuz I'm not entirely sure about this.

iobuhov replied:
Checked, no impact on old widgets

Keraito reviewed (changes requested):
Looks good, but some small changes are requested.

iobuhov commented on packages/pluggableWidgets/range-slider-web/src/RangeSlider.xml:
```diff
@@ -0,0 +1,202 @@
+                    <caption>Minimum value</caption>
+                    <description />
+                </property>
+                <property key="minAttribute" type="attribute" required="false">
```
I think yes, this is breaking change. Last time I was discussing this with @diego-antonelli, he told me to keep keys as much as I can.

Keraito replied:
Let's leave it then

iobuhov commented on packages/pluggableWidgets/range-slider-web/src/RangeSlider.xml:
```diff
@@ -0,0 +1,202 @@
+                </property>
+                <property key="tooltipTypeUpper" type="enumeration" defaultValue="value">
+                    <caption>Upper bound tooltip type</caption>
+                    <description>By default tooltip shows current value. Choose 'Custom' to create your own template.</description>
```
But both props in same place and In my view having two description is bit redundant.
If they will appear in two different places -- make sense, but in this case they one above other so...

Keraito replied:
I think it's better to repeat. Now one has a description while the other ones doesn't which can be confusing. Especially since the comment is valid for both. We also don't know how it will look in Studio and whether they'll stay close to each forever.

iobuhov replied:
Yeah, good point about Studio

iobuhov commented on packages/pluggableWidgets/range-slider-web/src/RangeSlider.xml:
```diff
@@ -0,0 +1,202 @@
+                <!-- Steps Size End -->
+            </propertyGroup>
+            <propertyGroup caption="Tooltip">
+                <property key="showTooltip" type="boolean" defaultValue="true">
+                    <caption>Show tooltip</caption>
+                    <description />
+                </property>
+                <property key="tooltipTypeLower" type="enumeration" defaultValue="value">
+                    <caption>Lower bound tooltip type</caption>
+                    <description />
+                    <enumerationValues>
+                        <enumerationValue key="value">Value</enumerationValue>
+                        <enumerationValue key="customText">Custom</enumerationValue>
+                    </enumerationValues>
+                </property>
+                <property key="tooltipLower" type="textTemplate" required="false">
+                    <caption>Tooltip</caption>
+                    <description />
+                </property>
+                <property key="tooltipTypeUpper" type="enumeration" defaultValue="value">
+                    <caption>Upper bound tooltip type</caption>
+                    <description>By default tooltip shows current value. Choose 'Custom' to create your own template.</description>
+                    <enumerationValues>
+                        <enumerationValue key="value">Value</enumerationValue>
+                        <enumerationValue key="customText">Custom</enumerationValue>
+                    </enumerationValues>
+                </property>
+                <property key="tooltipUpper" type="textTemplate" required="false">
+                    <caption>Tooltip</caption>
+                    <description />
+                </property>
```
Yes, this is true, but this division comes from the requirement that you can have "textTemplate" for tooltip. and because each "handle" has own value source, we have to use template for each tooltip individually.

Keraito replied:
I think you're explaining why we have the type attribute and the text template attribute, that's perfectly fine and I understand. My question is more that the design document from Chris didn't specify a tooltip for the lower bound and the upper bound, so I'm asking where this came from? Because AFAIK Slider also only has 1 tooltip for the current value. Did you add this yourself or discuss with Chris? @iobuhov

iobuhov replied:
I did it myself. This is just most simple solution to have tooltip for range slider. I expect get feedback on PM review

Keraito replied:
@cjfhodges can you make sure to keep this into account?

iobuhov commented on packages/pluggableWidgets/range-slider-web/src/components/Range.tsx:
```diff
@@ -0,0 +1,23 @@
+    rootStyle?: CSSProperties;
+}
+
+export const Range = ({ className, classNameSlider, rootStyle, ...rcRangeProps }: RangeProps): JSX.Element => (
```
Didn't knew that it should be named RangeSlider. I was thinking that everything in `src/components` have no strict naming conventions.

Keraito replied:
Nono it was a genuine quesiton 😄  There's no naming convention, but "Range" felt quite out if place and didn't add much value while I didn't see a reason to not use the widget name itself.

iobuhov commented on packages/pluggableWidgets/range-slider-web/src/utils/createHandleGenerator.tsx:
```diff
@@ -0,0 +1,60 @@
+import { DynamicValue } from "mendix";
+
+// Copied from https://github.com/react-component/slider/blob/8.6.6/src/Slider.jsx#L165
+interface HandleGeneratorProps {
+    className: string;
+    prefixCls?: string;
+    vertical: boolean;
+    offset: number;
+    value: number;
+    dragging?: boolean;
+    disabled?: boolean;
+    min?: number;
+    max?: number;
+    index: number;
+    tabIndex?: number;
+    style?: React.CSSProperties;
+    ref?: React.Ref<any>;
+}
```
```ts
    /**
     * A handle generator which could be used to customized handle.
     */
    handle?(props: any): React.ReactNode;
 ```
 
 As you can see they don't define type for handleGenerator fn, so what I did I wrote this types mannually.

Keraito replied:
But this type that you created is for the `Handle` component, which the library has typings available for:
`import { Handle, HandleProps } from "rc-slider";`
![image](https://user-images.githubusercontent.com/5955441/142995117-a892ad09-1e80-457a-98b6-8862a0b532d8.png)

```
interface HandleGeneratorProps extends HandleProps {
    dragging?: boolean;
    index: number;
}
```
Except for `dragging` and `index`, everything is used for the `Handle` component in your code. I don't quite understand why we can't do the above. Do you have a particular reason? I think that's better than copy pasting the types on our own.

iobuhov replied:
I didn't saw that HandleProps have same shape... I'll change it

Commit:
style(range-slider-web): apply requested changes

Commit:
fix(range-slider-web): apply requested changes

- inherit props from HandleProps
- add description for tooltipTypeUpper prop xml definition

Commit:
fix(range-slider-web): update tsconfig

Commit:
fix(range-slider-web): force range-slider to expand in flex container

Keraito commented on packages/pluggableWidgets/range-slider-web/src/ui/RangeSlider.scss:
```diff
@@ -1,6 +1,7 @@
 .widget-range-slider {
     padding: 4px 8px;
     margin-bottom: 16px;
+    flex-grow: 1;
```
@iobuhov can you communicate with @leonardomendix how to specifically test scenarios where this is relevant? Think it's good to make sure that this does what we want + doesn't interfere in cases where we don't need it.

leonardomendix committed:
test(range-slider-web): update atlas screenshot baseline
